be/src/exec/vectorized/partition/partition_hash_map.h

```diff
@@ -0,0 +1,288 @@
+                                           const NullableColumn* nullable_key_column, KeyLoader&& key_loader,
+                                           KeyAllocator&& key_allocator) {
+        if (nullable_key_column->only_null()) {
+            const auto size = chunk->num_rows();
+            if (null_key_value.chunks.empty() || null_key_value.remain_size <= 0) {
+                if (!null_key_value.chunks.empty()) {
+                    null_key_value.chunks.back()->append_selective(*chunk, null_key_value.select_indexes.data(), 0,
+                                                                   null_key_value.select_indexes.size());
+                }
+
+                alloc_new_buffer(null_key_value, chunk);
+            }
+            int32_t offset = 0;
+            auto cur_remain_size = size;
+            while (null_key_value.remain_size < cur_remain_size) {
+                null_key_value.chunks.back()->append(*chunk, offset, null_key_value.remain_size);
+                offset += null_key_value.remain_size;
+                cur_remain_size -= null_key_value.remain_size;
+
+                alloc_new_buffer(null_key_value, chunk);
+            }
+            null_key_value.chunks.back()->append(*chunk, offset, cur_remain_size);
+            null_key_value.remain_size = chunk_size - null_key_value.chunks.back()->num_rows();
+        } else {
+            const auto& null_flag_data = nullable_key_column->null_column()->get_data();
+            const auto size = chunk->num_rows();
+
+            for (uint32_t i = 0; i < size; i++) {
+                PartitionChunks* value_ptr = nullptr;
+                if (null_flag_data[i] == 1) {
+                    value_ptr = &null_key_value;
+                } else {
+                    const auto& key = key_loader(i);
+                    auto iter = hash_map.lazy_emplace(
+                            key, [&](const auto& ctor) { return ctor(key_allocator(key), PartitionChunks()); });
+                    value_ptr = &iter->second;
+                }
+
+                auto& value = *value_ptr;
+                if (value.chunks.empty() || value.remain_size <= 0) {
+                    if (!value.chunks.empty()) {
+                        value.chunks.back()->append_selective(*chunk, value.select_indexes.data(), 0,
+                                                              value.select_indexes.size());
+                    }
+
+                    alloc_new_buffer(value, chunk);
+                }
+                value.select_indexes.push_back(i);
+                value.remain_size--;
+            }
+
+            for (auto& [_, value] : hash_map) {
+                flush(value, chunk);
+            }
+            flush(null_key_value, chunk);
+        }
+    }
+};
+
+template <PrimitiveType primitive_type, typename HashMap>
+struct PartitionHashMapWithOneNumberKey : public PartitionHashMapBase {
```

Comment:
I think you could try only using nullable hash map for partition_hash_map, and do hash_null optimization for non-null data column.